samples/matter/lock/src/credentials/credentials_data_types.h

```diff
@@ -0,0 +1,424 @@
+	CHIP_ERROR ConvertToPlugin(EmberAfPluginDoorLockUserInfo &userInfo) const;
+
+	/**
+	 * @brief Serialize all fields into data buffer.
+	 *
+	 * `buffSize` must be set to at least the RequiredBufferSize size.
+	 *
+	 * @param buff buffer to store serialized data.
+	 * @param buffSize size of input buffer.
+	 * @return size_t output serialized data length, 0 value means error.
+	 */
+	size_t Serialize(void *buff, size_t buffSize);
+
+	/**
+	 * @brief Deserialize all fields from given buffer.
+	 *
+	 * `buffSize` must be set to at least the RequiredBufferSize size.
+	 *
+	 * @param buff buffer containing serialized data (by invoking serialized method).
+	 * @param buffSize size of output buffer.
+	 * @return CHIP_ERROR_BUFFER_TOO_SMALL if provided buffSize is too small.
+	 * @return CHIP_ERROR_INVALID_ARGUMENT if arguments are wrong.
+	 * @return CHIP_NO_ERROR if deserialization has been finished successfully.
+	 */
+	CHIP_ERROR Deserialize(const void *buff, size_t buffSize);
+};
+
+template <uint16_t N> struct IndexList {
+	struct __attribute__((__packed__)) List {
+		uint16_t mLength{ 0 };
+		uint16_t mIndexes[N] = { std::numeric_limits<uint16_t>::max() };
+	} mList;
+
+	/**
+	 * @brief Get required size for single IndexList entry.
+	 *
+	 * This method can be used to prepare a proper serialization buffer.
+	 *
+	 * @return size of single credential entry.
+	 */
+	constexpr static size_t RequiredBufferSize() { return sizeof(List); }
+
+	/**
+	 * @brief Serialize all fields into data buffer.
+	 *
+	 * `buffSize` must be set to at least the RequiredBufferSize size.
+	 *
+	 * @param buff buffer to store serialized data.
+	 * @param buffSize size of input buffer.
+	 * @return size_t output serialized data length, 0 value means error.
+	 */
+	size_t Serialize(void *buff, size_t buffSize);
+
+	/**
+	 * @brief Deserialize all fields from given buffer.
+	 *
+	 * `buffSize` must be set to at least the RequiredBufferSize size.
+	 *
+	 * @param buff buffer containing serialized data (by invoking serialized method).
+	 * @param buffSize size of output buffer.
+	 * @return CHIP_ERROR_BUFFER_TOO_SMALL if provided buffSize is too small.
+	 * @return CHIP_ERROR_INVALID_ARGUMENT if arguments are wrong.
+	 * @return CHIP_NO_ERROR if deserialization has been finished successfully.
+	 */
+	CHIP_ERROR Deserialize(const void *buff, size_t buffSize);
+
+	/**
+	 * @brief Add index to the list
+	 *
+	 * @param index index to be added.
+	 * @return CHIP_ERROR_BUFFER_TOO_SMALL if there is no space to add another index.
+	 * @return CHIP_ERROR_INVALID_ARGUMENT if the given `index` already exists in the container.
+	 * @return CHIP_NO_ERROR if the index has been added successfully.
+	 */
+	CHIP_ERROR AddIndex(uint16_t index);
+};
+
+template <uint16_t N> inline size_t IndexList<N>::Serialize(void *buff, size_t buffSize)
+{
+	if (!buff || buffSize < RequiredBufferSize()) {
+		return 0;
+	}
+
+	size_t offset = 0;
+	pack(buff, &mList.mLength, sizeof(mList.mLength), offset);
```

Comment:
Would it make sense to pack only `sizeof(mList.mIndexes[0]) * mList.mLength` instead of always serialize the full array?
Also, the length doesn't have to be serialized as this is stored separately by the storage subsystem, right?

Reply:
The second part means that you don't have to store the number of items explicitly. When you load this data, you will get the BLOB size, which you can divide by the item size and get the number of items that way. Not sure if that's OK from backwards compatibility but I was just pointing out that storing the length seems redundant.